rust/loqui_client/src/connection_handler.rs

```diff
@@ -0,0 +1,306 @@
+use crate::Config;
+use bytesize::ByteSize;
+use failure::{err_msg, Error};
+use futures::sync::oneshot::Sender as OneShotSender;
+use loqui_connection::handler::{DelegatedFrame, Handler, Ready, TransportOptions};
+use loqui_connection::{Encoder, FramedReaderWriter, IdSequence, LoquiError};
+use loqui_protocol::frames::{
+    Error as ErrorFrame, Frame, Hello, HelloAck, LoquiFrame, Push, Request, Response,
+};
+use loqui_protocol::upgrade::{Codec, UpgradeFrame};
+use loqui_protocol::VERSION;
+use loqui_protocol::{is_compressed, make_flags};
+use serde::{de::DeserializeOwned, Serialize};
+use std::collections::HashMap;
+use std::future::Future;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::await;
+use tokio::net::TcpStream;
+use tokio::prelude::*;
+use tokio_codec::Framed;
+
+#[derive(Debug)]
+pub enum InternalEvent<Encoded: Serialize + Send + Sync, Decoded: DeserializeOwned + Send + Sync> {
+    Request {
+        payload: Encoded,
+        waiter_tx: OneShotSender<Result<Decoded, Error>>,
+    },
+    Push {
+        payload: Encoded,
+    },
+}
+
+pub struct ConnectionHandler<E: Encoder> {
+    waiters: HashMap<u32, OneShotSender<Result<E::Decoded, Error>>>,
+    config: Arc<Config<E>>,
+}
+
+impl<E: Encoder> ConnectionHandler<E> {
+    pub fn new(config: Arc<Config<E>>) -> Self {
+        Self {
+            // TODO: should probably sweep these, probably request timeout
+            waiters: HashMap::new(),
+            config,
+        }
+    }
+}
+
+impl<E: Encoder> Handler for ConnectionHandler<E> {
+    type InternalEvent = InternalEvent<E::Encoded, E::Decoded>;
+    existential type UpgradeFuture: Send + Future<Output = Result<TcpStream, Error>>;
+    existential type HandshakeFuture: Send
+        + Future<
+            Output = Result<(Ready, FramedReaderWriter), (Error, Option<FramedReaderWriter>)>,
+        >;
+    existential type HandleFrameFuture: Send + Future<Output = Result<Response, (Error, u32)>>;
+
+    fn max_payload_size(&self) -> ByteSize {
+        self.config.max_payload_size
+    }
+
+    fn upgrade(&self, tcp_stream: TcpStream) -> Self::UpgradeFuture {
+        let max_payload_size = self.max_payload_size();
+        async move {
+            let framed_socket = Framed::new(tcp_stream, Codec::new(max_payload_size));
+            let (mut writer, mut reader) = framed_socket.split();
+            writer = match await!(writer.send(UpgradeFrame::Request)) {
+                Ok(writer) => writer,
+                Err(_e) => return Err(LoquiError::TcpStreamClosed.into()),
+            };
+            match await!(reader.next()) {
+                Some(Ok(UpgradeFrame::Response)) => Ok(writer.reunite(reader)?.into_inner()),
+                Some(Ok(UpgradeFrame::Request)) => Err(LoquiError::UpgradeFailed.into()),
+                Some(Err(e)) => {
+                    error!("Upgrade failed. error={:?}", e);
+                    Err(LoquiError::UpgradeFailed.into())
```

Comment:
Should we pass this inner error to the end user - rather than just dumping it in a log?